src/dataflow/source/kafka.rs

```diff
@@ -433,35 +338,34 @@ fn downgrade_capability(
     // in next_partition_ts
     if let Some(entries) = timestamp_histories.borrow_mut().get_mut(id) {
         for (pid, entries) in entries {
-            // Obtain the last offset processed (or 0 if no messages have yet been processed)
-            let last_offset = *last_processed_offset.get(pid).unwrap_or(&0);
+            let pid = match pid {
+                PartitionId::Kafka(pid) => *pid,
+                _ => unreachable!(
+                    "timestamp.rs should never send messages with non-Kafka partitions \
+                                   to Kafka sources."
+                ),
+            };
+            let last_offset = partition_metadata[pid as usize].0;
             // Check whether timestamps can be closed on this partition
             while let Some((partition_count, ts, offset)) = entries.first() {
                 assert!(
                     *offset >= start_offset,
                     "Internal error! Timestamping offset went below start: {} < {}. Materialize will now crash.",
                     offset, start_offset
                 );
-                if partition_count > current_partition_count {
+                if *partition_count as usize > partition_metadata.len() {
                     // A new partition has been added, we need to update the appropriate
-                    // entries before we continue. This will also update the last_processed_offset
-                    // and next_partition_ts data structures
-                    let partitions = update_partition_list(
-                        consumer,
-                        topic,
-                        *partition_count,
-                        last_processed_offset,
-                        next_partition_ts,
-                        *last_closed_ts,
-                        start_offset,
+                    // entries before we continue.
+                    partition_metadata.extend(
```

Comment:
I don't think this is correct. This should be last_closed_offset. When you add a new partition, you have to add it with a timestamp that is strictly greater than the last_closed_timestamp. Otherwise you'd end up in a situation where you go back in time. That was the use for last_closed_ts. start_offset is not enough.

Reply:
Oh, you're right